optaplanner-core/src/main/java/org/optaplanner/core/impl/util/ConsecutiveSetTree.java

```diff
@@ -0,0 +1,314 @@
+package org.optaplanner.core.impl.util;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ConsecutiveSetTree<T, I extends Comparable<I>, D extends Comparable<D>> implements Collection<T> {
+    Function<T, I> indexFunction;
+    BiFunction<I, I, D> differenceFunction;
+    D maxDifference;
+    D zeroDifference;
+    TreeMap<T, Sequence<T>> startItemToSequence;
+    Class<? extends T> typeClass;
+    Comparator<T> comparator;
+
+    TreeMapValueList<T, Sequence<T>> sequenceList;
+    ConsecutiveData<T, D> consecutiveData;
+
+    public ConsecutiveSetTree(Class<? extends T> typeClass, Function<T, I> indexFunction,
+            BiFunction<I, I, D> differenceFunction, D maxDifference, D zeroDifference) {
+        this.indexFunction = indexFunction;
+        this.differenceFunction = differenceFunction;
+        this.maxDifference = maxDifference;
+        this.zeroDifference = zeroDifference;
+        this.typeClass = typeClass;
+        // Hashcode for duplicate protection
+        // Ex: two different games on the same time slot
+        comparator = Comparator.comparing(indexFunction).thenComparingInt(System::identityHashCode);
+        startItemToSequence = new TreeMap<>(comparator);
+        consecutiveData = new ConsecutiveData<>(this);
+        sequenceList = new TreeMapValueList<>(startItemToSequence);
+    }
+
+    public List<Sequence<T>> getConsecutiveSequences() {
+        return sequenceList;
+    }
+
+    public List<D> getBreaks() {
+        return startItemToSequence.keySet().stream().flatMap(startItem -> {
+            T nextStartItem = startItemToSequence.higherKey(startItem);
+            if (nextStartItem == null) {
+                return Stream.empty();
+            }
+            return Stream.of(differenceFunction.apply(indexFunction.apply(startItemToSequence.get(startItem).getItems().last()),
+                    indexFunction.apply(nextStartItem)));
+        }).collect(Collectors.toList());
+    }
+
+    public List<D> getConsecutiveLengths() {
+        return getConsecutiveSequences().stream()
+                .map(sequence -> differenceFunction.apply(indexFunction.apply(sequence.getItems().first()),
+                        indexFunction.apply(sequence.getItems().last())))
+                .collect(Collectors.toList());
+    }
```

Comment:
Do we need this? Where?
If we need this, it must be fully incremental. If we don't need this now, I suggest we remove it for now.